### PR TITLE
 LTD-3654-pull-mentions-data 

### PR DESCRIPTION
**Pull request by depsiatwal**
https://uktrade.atlassian.net/browse/LTD-3654

This API call has been linked with the notes and timelines screen. 

The mentions portion will pull the mentions on a screen and display as per requirements. We could have used the same data shape as the Audit Table but I believe this is fundamentally wrong and a bit of a hack. Hence a new API call to pull real data from the source table.

### CODE REVIEW

**Review by Tllew — APPROVED** [review]
Looks good

File: api/cases/serializers.py [hunk 1]
```diff
@@ -361,6 +361,11 @@ class CaseNoteMentionsSerializer(serializers.ModelSerializer):
     user = PrimaryKeyRelatedSerializerField(
         queryset=GovUser.objects.all(), serializer=GovUserViewSerializer, required=False
     )
+    case_note_user = PrimaryKeyRelatedSerializerField(
+        queryset=BaseUser.objects.all(), source="case_note.user", serializer=BaseUserViewSerializer, required=False
```
Comment (saruniitr) [inline api/cases/serializers.py:365]: best to limit this to `GovUser` only

File: api/cases/tests/test_case_notes.py [hunk 2]
```diff
@@ -167,3 +166,22 @@ def test_view_pre_submitted_case_notes_post_submit(self):
...
+
+        case_note = self.create_case_note(self.case, "Hairpin Turns", self.gov_user.baseuser_ptr)
+        self.other_user = self.create_gov_user("test@gmail.com", self.team)  # /PS-IGNORE
+        self.create_case_note_mention(case_note, self.other_user)
+        response = self.client.get(self.url, **self.gov_headers)
+        result = response.json()
+
+        self.assertEqual(result["count"], 1)
+        self.assertEqual(result["results"][0]["user"]["id"], str(self.other_user.baseuser_ptr.id))
+        self.assertEqual(result["results"][0]["case_note_user"]["id"], str(self.gov_user.baseuser_ptr.id))
```
Comment (currycoder) [inline api/cases/tests/test_case_notes.py:187]: Would suggest asserting the whole result e.g. `assert result == expected_result`.  Keeps the tests tight and exhaustive

File: api/cases/views/views.py [hunk 3]
```diff
@@ -1162,3 +1164,11 @@ def get_queryset(self):
                 "control_list_entries",
             )
         )
+
+
+class CaseNoteMentionList(ListAPIView):
```
Comment (currycoder) [inline api/cases/views/views.py:1169]: I wonder if this should all be in its own views file - this one is getting pretty chonky with a lot of unrelated stuff.

File: api/cases/serializers.py (hunk 1 above)
Comment (currycoder) [reply api/cases/serializers.py:365]: I don't really know what `PrimaryKeyRelatedSerializerField` gets us; I think replacing with the following gives us more options for selecting all the data in a single query;
```
    case_note_user = BaseUserViewSerializer(
        source="case_note.user", required=False
    )
```

File: api/cases/serializers.py [hunk 4]
```diff
@@ -361,6 +361,11 @@ class CaseNoteMentionsSerializer(serializers.ModelSerializer):
     user = PrimaryKeyRelatedSerializerField(
```
Comment (currycoder) [inline api/cases/serializers.py:361]: Same as below, maybe replace with;
```
    user = GovUserViewSerializer()
```

File: api/cases/views/views.py [hunk 5]
```diff
@@ -1162,3 +1164,11 @@ def get_queryset(self):
...
+    authentication_classes = (GovAuthentication,)
+    serializer_class = CaseNoteMentionsSerializer
+
+    def get_queryset(self):
+        return CaseNoteMentions.objects.filter(case_note__case_id=self.kwargs["pk"]).order_by("-created_at")
```
Comment (currycoder) [inline api/cases/views/views.py:1174]: ATM, I think the serializer will perform three extra queries per CaseNoteMention record to populate `case_note`, `user`, `case_note.user`.  I think this could be avoided with;
```
CaseNoteMentions.objects
    .select_related("case_note", "user", "case_note__user")    
    .filter(case_note__case_id=self.kwargs["pk"]).order_by("-created_at")
```

File: api/cases/views/views.py [hunk 6]
```diff
@@ -1162,3 +1164,11 @@ def get_queryset(self):
                 "control_list_entries",
             )
         )
+
+
+class CaseNoteMentionList(ListAPIView):
+    authentication_classes = (GovAuthentication,)
+    serializer_class = CaseNoteMentionsSerializer
+
+    def get_queryset(self):
```
Comment (currycoder) [inline api/cases/views/views.py:1173]: Functionality wise, it seems like this pulls all CaseNoteMention objects for the case - is there a need to get mentions just for a single user?

File: api/cases/serializers.py (hunk 1 above)
Comment (depsiatwal) [reply api/cases/serializers.py:365]: > best to limit this to `GovUser` only

I was forced to use baseUser as it seems to have issues serialising Team field. 
I think it's save for now since this is used for relationships as opposed to creating.

Comment (depsiatwal) [reply api/cases/serializers.py:365]: > I don't really know what `PrimaryKeyRelatedSerializerField` gets us; I think replacing with the following gives us more options for selecting all the data in a single query;
> 
> ```
>     case_note_user = BaseUserViewSerializer(
>         source="case_note.user", required=False
>     )
> ```
PrimaryKeyRelatedField this allows a list of Id's to be passed to create a relationship between the tables. 
else it would expected the user object to be created and expects a dict.

File: api/cases/tests/test_case_notes.py (hunk 2 above)
Comment (depsiatwal) [reply api/cases/tests/test_case_notes.py]: yeah I also prefer assert result == expected_result
was just trying to keep it consistent  with the other tests 

anyway changed ...

File: api/cases/views/views.py (hunk 5 above)
Comment (depsiatwal) [reply api/cases/views/views.py]: ahh nice one changed

File: api/cases/views/views.py (hunk 6 above)
Comment (depsiatwal) [reply api/cases/views/views.py]: there will be that I think will require a very different view of the data 
will consolidate and refactor is need be

File: api/cases/tests/test_case_notes.py (hunk 2 above)
Comment (currycoder) [reply api/cases/tests/test_case_notes.py]: Sorry my point wasn't about `self.assertEqual` vs `assert` but about asserting that the *whole* result JSON was what we expected it to be rather than just individual fields.  IMO that gives us proper coverage of the format of the API.

**Review by currycoder — APPROVED** [review]
Nice one! Just one comment left regarding testing result formats which I will leave with you.

File: api/cases/tests/test_case_notes.py [hunk 7]
```diff
@@ -30,21 +29,27 @@ def test_create_case_note_with_mentions_successful(self):
...
-        self.assertEqual(CaseNote.objects.get().is_urgent, True)
-        self.assertEqual(CaseNoteMentions.objects.count(), 1)
+        assert response.status_code == status.HTTP_201_CREATED
+        assert CaseNote.objects.count() == 1
+        assert CaseNote.objects.get().text == self.data.get("text")
+        assert CaseNote.objects.get().is_visible_to_exporter is False
+        assert CaseNote.objects.get().is_urgent is True
+        assert CaseNoteMentions.objects.count() == 2
         assert response.json()["case_note"]["mentions"][0]["user"]["id"] == mentions[0]["user"]
+        assert response.json()["case_note"]["mentions"][1]["user"]["id"] == mentions[1]["user"]
```
Comment (kevincarrogan) [inline api/cases/tests/test_case_notes.py:52]: We use `self.assertX` for the API not `assert`.
